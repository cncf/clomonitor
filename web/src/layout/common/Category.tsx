import classNames from 'classnames';
import { isUndefined } from 'lodash';

import getCategoryColor from '../../utils/getCategoryColor';
import styles from './Category.module.css';

interface Props {
  icon?: JSX.Element;
  value: number;
  name: string;
  shortName?: string;
  bigSize?: boolean;
  colNumber?: number;
}

const Category = (props: Props) => {
  const color = getCategoryColor(props.value);
  return (
    <div className={`col-${props.colNumber || 6} text-truncate`}>
      <small className={`text-muted fw-bold text-nowrap ${styles.title}`}>
        {props.icon && <span className={`pe-1 d-inline-block position-relative ${styles.icon}`}>{props.icon}</span>}
        <span
          className={classNames({
            'd-none d-md-inline-block': !isUndefined(props.shortName),
          })}
        >
          {props.name}
        </span>
        {!isUndefined(props.shortName) && <span className="d-inline-block d-md-none">{props.shortName}</span>}
      </small>
      <div className={`d-flex flex-row bg-white position-relative border overflow-hidden ${styles.line}`}>
        <div className={`text-center fw-bold font-monospace ${styles.value} ${props.bigSize ? styles.bigSize : ''}`}>
          {props.value}
        </div>
        <div className="flex-grow-1 w-100 position-relative">
          <div
            className="position-absolute top-0 start-0 bottom-0"
            style={{
              width: props.value === 100 ? '100%' : `calc(${props.value}% - 5px)`,
              backgroundColor: `var(--rm-${color})`,
            }}
          >
            {props.value !== 100 && (
              <div
                style={{ borderLeftColor: `var(--rm-${color})` }}
                className={`position-absolute ${styles.arrow} ${props.bigSize ? styles.bigArrow : ''}`}
              />
            )}
          </div>
        </div>
      </div>
    </div>
  );
};

export default Category;
