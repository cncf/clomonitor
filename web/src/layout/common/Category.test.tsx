import { render, screen } from '@testing-library/react';

import Category from './Category';

const defaultProps = {
  value: 80,
  name: 'Documentation',
  shortName: 'Docs',
};

describe('Category', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(<Category {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders proper content', () => {
    render(<Category {...defaultProps} />);
    expect(screen.getByText('Documentation')).toBeInTheDocument();
    expect(screen.getByText('Documentation')).toHaveClass('d-none d-md-inline-block');
    expect(screen.getByText('Docs')).toBeInTheDocument();
    expect(screen.getByText('Docs')).toHaveClass('d-inline-block d-md-none');
    expect(screen.getByText('80')).toBeInTheDocument();

    const line = screen.getByTestId('line');
    expect(line).toBeInTheDocument();
    expect(line).toHaveStyle('width: calc(80% - 5px)');

    expect(screen.getByTestId('peak')).toBeInTheDocument();
  });

  it('does not render peak when value is 100', () => {
    render(<Category {...defaultProps} value={100} />);
    expect(screen.queryByTestId('peak')).toBeNull();
  });
});
