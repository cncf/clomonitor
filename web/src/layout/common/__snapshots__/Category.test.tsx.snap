// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="col-6 text-truncate"
  >
    <div
      class="d-flex flex-row align-items-baseline text-muted fw-bold flex-nowrap title"
    >
      <span
        class="text-truncate d-none d-md-inline-block"
      >
        Documentation
      </span>
      <span
        class="d-inline-block d-md-none text-truncate"
      >
        Docs
      </span>
    </div>
    <div
      class="d-flex flex-row bg-white position-relative border overflow-hidden line"
    >
      <div
        class="text-center fw-bold font-monospace value "
      >
        80
      </div>
      <div
        class="flex-grow-1 w-100 position-relative"
      >
        <div
          class="position-absolute top-0 start-0 bottom-0"
          data-testid="line"
          style="width: calc(80% - 5px);"
        >
          <div
            class="position-absolute arrow "
            data-testid="peak"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
